File: js/tests/sunspider-1.0/c,c++/access-fannkuch.c
/* The Great Computer Language Shootout
   http://shootout.alioth.debian.org/
   contributed by Isaac Gouy */

int fannkuch(int n) {
    int check = 0;
    int perm[n];
    int perm1[n];
    int count[n];
    int maxPerm[n];
    int maxFlipsCount = 0;
    int m = n - 1;
    int i;

    for (i = 0; i < n; i++) {
        perm1[i] = i;
    }
    int r = n;

    while (1) {

        while (r != 1) {
             count[r - 1] = r;
             r--;
        }
        if (!(perm1[0] == 0 || perm1[m] == m)) {
            int i;
            for (i = 0; i < n; i++) {
                perm[i] = perm1[i];
                }

            int flipsCount = 0;
            int k;

            while (!((k = perm[0]) == 0)) {
            int k2 = (k + 1) >> 1;
            for (i = 0; i < k2; i++) {
                int temp = perm[i]; perm[i] = perm[k - i]; perm[k - i] = temp;
            }
            flipsCount++;
            }

            if (flipsCount > maxFlipsCount) {
            maxFlipsCount = flipsCount;
            for (int i = 0; i < n; i++) {
                maxPerm[i] = perm1[i];
            }
            }
        }

        while (1) {
            if (r == n) {
                return maxFlipsCount;
            }
            int perm0 = perm1[0];
            int i = 0;
            while (i < r) {
                int j = i + 1;
                perm1[i] = perm1[j];
                i = j;
            }
            perm1[r] = perm0;

            count[r] = count[r] - 1;
            if (count[r] > 0) {
                break;
            }
            r++;
        }
    }
} 


int main() {
    return fannkuch(10);
}

File: js/tests/sunspider-1.0/ts/access-fannkuch.ts
/* The Great Computer Language Shootout
   http://shootout.alioth.debian.org/
   contributed by Isaac Gouy */

function fannkuch(n: int): int {
   let check: int = 0;
   var perm: int[] = new Array(n);
   var perm1: int[] = new Array(n);
   var count: int[] = new Array(n);
   var maxPerm: int[] = new Array(n);
   var maxFlipsCount: int = 0;
   var m: int = n - 1;

   for (let i: int = 0; i < n; i++){
       perm1[i] = i;
   }
   var r: int = n;

   while (true) {

      while (r != 1) {
           count[r - 1] = r;
            r--;
      }
      if (!(perm1[0] == 0 || perm1[m] == m)) {
         for (let i: int = 0; i < n; i++){
            perm[i] = perm1[i];
         } 

         var flipsCount: int = 0;
         var k: int;

         while (!((k = perm[0]) == 0)) {
            var k2: int = (k + 1) >> 1;
            for (let i: int = 0; i < k2; i++) {
               var temp: int = perm[i];
                perm[i] = perm[k - i]; 
                perm[k - i] = temp;
            }
            flipsCount++;
         }

         if (flipsCount > maxFlipsCount) {
            maxFlipsCount = flipsCount;
            for (let i: int = 0; i < n; i++){
                maxPerm[i] = perm1[i];
            } 
         }
      }

      while (true) {
         if (r == n){
            return maxFlipsCount;
         } 
         var perm0: int = perm1[0];
         let i: int = 0;
         while (i < r) {
            let j: int = i + 1;
            perm1[i] = perm1[j];
            i = j;
         }
         perm1[r] = perm0;

         count[r] = count[r] - 1;
         if (count[r] > 0){
            break;
         }
         r++;
      }
   }
}

export function main(): int{
    return fannkuch(10); //CHANGED THIS CAUSE 8 WAS TOO FAST // returns 38
}

